validator_worker/src/sentry_interface.rs

```diff
@@ -1,143 +1,264 @@
 use futures::Future;
 use futures_legacy::Future as LegacyFuture;
+use primitives::adapter::Adapter;
+use primitives::sentry::{
+    ChannelAllResponse, EventAggregateResponse, LastApprovedResponse, SuccessResponse,
+    ValidatorMessageResponse,
+};
+use primitives::validator::MessageTypes;
+use primitives::{Channel, Config, ValidatorDesc};
+use reqwest::header::AUTHORIZATION;
 use reqwest::r#async::{Client, Response};
-use reqwest::Error;
 use serde::Deserialize;
+use std::error::Error;
 use std::iter::once;
+use std::pin::Pin;
+use std::time::Duration;
 
-
-pub fn all_channels(
-    &self,
-    validator: Option<&ValidatorId>,
-) -> impl Future<Output = Result<Vec<Channel>, Error>> {
-    let validator = validator.cloned();
-    // call Sentry and fetch first page, where validator = identity
-    let first_page = self.clone().fetch_page(1, validator.clone());
-
-    let handle = self.clone();
-    first_page
-        .and_then(move |response| {
-            let first_page_future = ok(response.channels).boxed();
-
-            if response.total_pages < 2 {
-                // if there is only 1 page, return the results
-                first_page_future
-            } else {
-                // call Sentry again for the rest of tha pages
-                let futures = (2..=response.total_pages)
-                    .map(|page| {
-                        handle
-                            .clone()
-                            .fetch_page(page, validator.clone())
-                            .map(|response_result| {
-                                response_result.and_then(|response| Ok(response.channels))
-                            })
-                            .boxed()
-                    })
-                    .chain(once(first_page_future));
-
-                try_join_all(futures)
-                    .map(|result_all| {
-                        result_all
-                            .and_then(|all| Ok(all.into_iter().flatten().collect::<Vec<_>>()))
-                    })
-                    .boxed()
-            }
-        })
-        .boxed()
-}
-
-async fn fetch_page(
-    self,
-    page: u64,
-    validator: Option<ValidatorId>,
-) -> Result<ChannelAllResponse, reqwest::Error> {
-    let mut query = vec![format!("page={}", page)];
-
-    if let Some(validator) = validator {
-        query.push(format!("validator={}", validator));
-    }
-
-    let future = self
-        .client
-        .get(format!("{}/channel/list?{}", self.sentry_url, query.join("&")).as_str())
-        .send()
-        .and_then(|mut res: Response| res.json::<ChannelAllResponse>());
-
-    await!(future.compat())
-}
-
-
-#[derive(Clone)]
-pub(crate) struct SentryApi {
-    pub sentry_url: String,
+#[derive(Clone, Debug)]
+pub struct SentryApi<T: Adapter> {
+    pub adapter: T,
+    pub validator_url: String,
     pub client: Client,
     pub logging: bool,
-    pub channel: Channel
+    pub channel: Channel,
```

Comment:
I'm still not sure why won't you pass the channel to the functions that need it, instead of having it on instantiation of the struct.
E.g. in the methods: get_latest_msg, get_last_msgs, propagate and etc. you don't need it. They need identity - they have it. The functions requiring the channel to have validator with `whoami` value should handle this on their own IMO. Otherwise we are blocking these generic non-channel dependant functions to require a channel. At the end we might end-up using an instance of 1 SentryApi with 1 channel to do work for totally different channel(for example batch processing).

Reply:
You have a point.

We should rethink the Sentry API interface; perhaps channel should not be used in instantiation, but rather for the methods that need it

Reply:
The channel is needed in instantiation to get the channel validator url and validate that the validator is meant to be validating the channel.
Also depending on how the sentry interface grows there could definitely be a need for other channel properties in the future.
Also, Sentry API is channel dependent ( i.e. a Sentry API is for one channel, so for multiple channels, there would be multiple Sentry API's )

Reply:
I guess there are multiple way to work on this issue indeed.
We might not check if the identity is in the channel inside `new`, but instead where we work with the channel. The rest of the methods don't deal with the Channel either way and if you try to get messages for an identity that doesn't exist - that's ok, right?

We might separate this? Doing fetching of aggregates, channels, messages + posting some messages and etc. isn't it too much for 1 object?
I am trying to also figure that out myself.

Reply:
@elpiel There is no need for separating this. They all point to the same resource.
Sentry API is channel dependent, each of those methods actually uses the channel in an abstracted way i.e. validator_url

Reply:
So let me get this straight, you want the `Heartbeat`/`producer`/`leader`/`follower` fnunctions to get a `SentryApi` instance which is bound to a channel and instantiate it before calling those functions?

Reply:
I think you should review the JS implementation. The sentry API is channel dependent and so is the heartbeat, producer, leader, follower

Reply:
Yeah, but I mean if you have the Channel in the SentryApi, then you don't need to pass it to any of the heartbeat, producer, leader, follower, you only need SentryApi